File: frontend/src/api/clean.ts
import { API_URL, ApiError, request } from "./http";

// ─── Types ──────────────────────────────────────────────────────────────────

export interface CriteriaCount {
  selected_users: number;
  threshold_chars?: number;
  threshold_seconds?: number;
}

export interface CentralMeasures {
  mean: number;
  mode: number;
  median: number;
  iqr_lower: number;
  iqr_upper: number;
}

export interface PreviewResponse {
  collection_id: string;
  total_users: number;
  central_measures: CentralMeasures;
  by_criteria: Record<string, CriteriaCount>;
  union_if_combined: number;
}

export interface DatasetResponse {
  dataset_id: string;
  name: string;
  collection_id: string;
  video_id: string;
  total_users_original: number;
  total_users_selected: number;
  criteria_applied: string[];
  created_at: string;
}

export interface DatasetSummary {
  dataset_id: string;
  name: string;
  video_id: string;
  total_users_selected: number;
  criteria_applied: string[];
  created_at: string;
}

export interface DatasetCreateRequest {
  collection_id: string;
  criteria: string[];
  thresholds?: {
    threshold_chars?: number;
    threshold_seconds?: number;
  };
}

// ─── API ────────────────────────────────────────────────────────────────────

export const cleanApi = {
  preview: (
    params: {
      collection_id: string;
      criteria: string;
      threshold_chars?: number;
      threshold_seconds?: number;
    },
    token: string
  ) => {
    const qs = new URLSearchParams();
    qs.set("collection_id", params.collection_id);
    qs.set("criteria", params.criteria);
    if (params.threshold_chars != null) qs.set("threshold_chars", String(params.threshold_chars));
    if (params.threshold_seconds != null)
      qs.set("threshold_seconds", String(params.threshold_seconds));

    return request<PreviewResponse>(`/clean/preview?${qs}`, {}, token);
  },

  create: (data: DatasetCreateRequest, token: string) =>
    request<DatasetResponse>("/clean", { method: "POST", body: JSON.stringify(data) }, token),

  list: (token: string, videoId?: string) => {
    const qs = videoId ? `?video_id=${videoId}` : "";
    return request<DatasetSummary[]>(`/clean/datasets${qs}`, {}, token);
  },

  download: async (
    datasetId: string,
    format: "json" | "csv",
    token: string,
    datasetName: string
  ): Promise<void> => {
    const res = await fetch(`${API_URL}/clean/datasets/${datasetId}/download?format=${format}`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (!res.ok) {
      const body = (await res.json().catch(() => ({}))) as { detail?: string };
      throw new ApiError(body.detail ?? "Erro ao baixar dataset.", res.status);
    }
    const blob = await res.blob();
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = `${datasetName}.${format}`;
    a.click();
    URL.revokeObjectURL(url);
  },

  delete: (datasetId: string, token: string) =>
    request<void>(`/clean/datasets/${datasetId}`, { method: "DELETE" }, token),

  import: async (
    data: {
      dataset: { name: string; video_id: string; criteria_applied: string[] };
      users: Array<{
        author_channel_id: string;
        author_display_name: string;
        comment_count: number;
        matched_criteria: string[];
      }>;
    },
    token: string,
    onProgress?: (sent: number, total: number) => void
  ): Promise<DatasetResponse> => {
    const CHUNK_SIZE = 2000;
    const allUsers = data.users;
    const total = allUsers.length;
    const firstBatch = allUsers.slice(0, CHUNK_SIZE);
    const hasMore = total > CHUNK_SIZE;

    const result = await request<DatasetResponse>(
      "/clean/import",
      {
        method: "POST",
        body: JSON.stringify({
          dataset: data.dataset,
          users: firstBatch,
          done: !hasMore,
        }),
      },
      token
    );
    onProgress?.(firstBatch.length, total);

    if (!hasMore) return result;

    let offset = CHUNK_SIZE;
    while (offset < total) {
      const chunk = allUsers.slice(offset, offset + CHUNK_SIZE);
      const isLast = offset + chunk.length >= total;
      await request<{
        dataset_id: string;
        total_users: number;
        chunk_received: number;
        done: boolean;
      }>(
        "/clean/import-chunk",
        {
          method: "POST",
          body: JSON.stringify({
            dataset_id: result.dataset_id,
            users: chunk,
            done: isLast,
          }),
        },
        token
      );
      offset += chunk.length;
      onProgress?.(offset, total);
    }

    return result;
  },
};

File: backend/schemas/clean.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

VALID_CRITERIA = Literal[
    "percentil",
    "media",
    "moda",
    "mediana",
    "curtos",
    "intervalo",
    "identicos",
    "perfil",
]


class CleanThresholds(BaseModel):
    threshold_chars: int = Field(default=20, ge=1, le=500)
    threshold_seconds: int = Field(default=30, ge=1, le=3600)


# ─── Request ─────────────────────────────────────────────────────────────────


class DatasetCreate(BaseModel):
    collection_id: uuid.UUID
    criteria: list[VALID_CRITERIA] = Field(min_length=1)
    thresholds: CleanThresholds = CleanThresholds()


class DatasetImportUser(BaseModel):
    author_channel_id: str = Field(min_length=1)
    author_display_name: str = ""
    comment_count: int = Field(default=0, ge=0)
    matched_criteria: list[str] = Field(default_factory=list)


class DatasetImportMeta(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    video_id: str = Field(default="", min_length=0)
    criteria_applied: list[str] = Field(default_factory=list)
    total_users_original: int | None = None
    total_users_selected: int | None = None


class DatasetImport(BaseModel):
    """Aceita o mesmo formato JSON exportado pelo download."""

    dataset: DatasetImportMeta
    users: list[DatasetImportUser] = Field(min_length=1)
    comments: list | None = None  # ignorado no import — comments já estão na coleta


# ─── Response — Preview ──────────────────────────────────────────────────────


class CriteriaCount(BaseModel):
    selected_users: int
    threshold_chars: int | None = None
    threshold_seconds: int | None = None


class CentralMeasures(BaseModel):
    mean: float
    mode: float
    median: float
    iqr_lower: float
    iqr_upper: float


class PreviewResponse(BaseModel):
    collection_id: uuid.UUID
    total_users: int
    central_measures: CentralMeasures
    by_criteria: dict[str, CriteriaCount]
    union_if_combined: int


# ─── Response — Dataset ──────────────────────────────────────────────────────


class DatasetResponse(BaseModel):
    dataset_id: uuid.UUID
    name: str
    collection_id: uuid.UUID
    video_id: str
    total_users_original: int
    total_users_selected: int
    criteria_applied: list[str]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DatasetSummary(BaseModel):
    dataset_id: uuid.UUID
    name: str
    video_id: str
    total_users_selected: int
    criteria_applied: list[str]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)
